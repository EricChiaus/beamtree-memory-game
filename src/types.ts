export interface FlippableObject {
    isFlippable: boolean;
    isFlipped: boolean;
}

export interface FlippableCard extends FlippableObject {
    index: number;
    content: string;
}

export interface Game<T> {
    items: T[];
    isDisabled: boolean;
    hasCompleted: boolean;
    restart: () => void;
    flip: (index: number) => void;
    getItem: (index: number) => T | null;
}
