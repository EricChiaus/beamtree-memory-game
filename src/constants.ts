export const BOARD_WIDTH = 4;
export const BOARD_HEIGHT = 4;
export const CARDS_NUMBER = BOARD_WIDTH * BOARD_HEIGHT;
