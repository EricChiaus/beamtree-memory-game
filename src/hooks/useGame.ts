import { useEffect, useState, useCallback, useMemo, useRef } from 'react';
import { FlippableCard, Game } from './../types';
import { flipCard, generateCards } from './../helpers/cards';

const useGame = (): Game<FlippableCard> => {
    const [items, setItems] = useState<FlippableCard[]>([]);
    const [isDisabled, setIsDisabled] = useState<boolean>(false);
    const last = useRef<FlippableCard | null>(null);

    const hasCompleted = useMemo<boolean>(
        () => items.every(({ isFlipped }) => !isFlipped),
        [items],
    );

    const restart = useCallback(() => {
        setItems(generateCards());
    }, []);

    const getItem = useCallback((index: number) => items[index], [items]);

    const flip = useCallback(
        (index: number) => {
            if (isDisabled || !items[index].isFlippable) {
                return;
            }
            setItems([...items]);
            items[index] = flipCard(items[index]);
            if (last.current) {
                if (!items[index].isFlipped) {
                    if (last.current.content === items[index].content) {
                        // Two cards flipped to front side with same content, mark them unflippable
                        items[last.current.index] = {
                            ...items[last.current.index],
                            isFlippable: false,
                        };
                        items[index] = { ...items[index], isFlippable: false };
                        setItems([...items]);
                    } else {
                        // Not same content, set a timer to flip them back
                        items[last.current.index] = {
                            ...items[last.current.index],
                            isFlipped: true,
                        };
                        items[index] = { ...items[index], isFlipped: true };
                        setIsDisabled(true);
                        setTimeout(() => {
                            setItems([...items]);
                            setIsDisabled(false);
                        }, 1000);
                    }
                }
                last.current = null;
            } else {
                // No card has been flipped yet, save this one.
                last.current = items[index];
            }
        },
        [isDisabled, items],
    );

    useEffect(restart, [restart]);

    return {
        items,
        isDisabled,
        hasCompleted,
        restart,
        flip,
        getItem,
    };
};

export default useGame;
