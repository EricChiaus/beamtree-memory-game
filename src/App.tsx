import React from 'react';
import { Button, Container, WinText } from './App.styles';
import GameContext from './contexts/GameContext';
import useGame from './hooks/useGame';
import Board from './components/Board';

const App: React.FC = () => {
    const game = useGame();

    return (
        <GameContext.Provider value={game}>
            <Container>
                <Board />
                {game.hasCompleted ? (
                    <WinText>Congrats you did it!</WinText>
                ) : null}
                <Button onClick={game.restart}>Restart</Button>
            </Container>
        </GameContext.Provider>
    );
};

export default App;
