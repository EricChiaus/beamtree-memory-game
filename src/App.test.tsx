import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';
import useGame from './hooks/useGame';
import { CARDS_NUMBER } from './constants';
import { FlippableCard } from './types';

const generateCards = (): FlippableCard[] => {
    const cards: FlippableCard[] = [];
    for (let i = 0; i < CARDS_NUMBER; i++) {
        cards.push({
            index: i,
            content: Math.floor(i / 2).toString(),
            isFlippable: true,
            isFlipped: true,
        });
    }
    return cards;
};

jest.mock('./hooks/useGame', () => jest.fn());
const mockUseGame = useGame as jest.Mock;

const restartSpy = jest.fn();

describe('<App />', () => {
    it('should match the snapshot', () => {
        mockUseGame.mockReset().mockReturnValue({
            items: generateCards(),
            isDisabled: false,
            hasCompleted: false,
            restart: jest.fn(),
            flip: jest.fn(),
            getItem: jest.fn(),
        });
        const { container } = render(<App />);
        expect(container).toMatchSnapshot();
    });

    describe('when game is completed', () => {
        it('should render congrats text', () => {
            mockUseGame.mockReset().mockReturnValue({
                items: generateCards(),
                isDisabled: false,
                hasCompleted: true,
                restart: jest.fn(),
                flip: jest.fn(),
                getItem: jest.fn(),
            });
            render(<App />);
            expect(
                screen.getByText('Congrats you did it!'),
            ).toBeInTheDocument();
        });
    });

    describe('when clicks restart button', () => {
        it('should invoke restart event', () => {
            mockUseGame.mockReset().mockReturnValue({
                items: generateCards(),
                isDisabled: false,
                hasCompleted: true,
                restart: restartSpy,
                flip: jest.fn(),
                getItem: jest.fn(),
            });
            render(<App />);
            const button = screen.getByRole('button', { name: 'Restart' });
            expect(button).toBeInTheDocument();
            fireEvent.click(button);
            expect(restartSpy).toBeCalledTimes(1);
        });
    });
});
