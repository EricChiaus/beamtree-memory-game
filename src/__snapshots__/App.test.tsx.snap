// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `
<div>
  <div
    class="sc-bcXHqe jxqCzq"
  >
    <div
      class="sc-hLBbgP beMrxu"
    >
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-0"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-1"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-2"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-3"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-4"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-5"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-6"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-7"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-8"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-9"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-10"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-11"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-12"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-13"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-14"
      />
      <button
        class="sc-eDvSVe eicupT"
        data-testid="flippable-card-15"
      />
    </div>
    <button
      class="sc-gswNZR kCorOY"
    >
      Restart
    </button>
  </div>
</div>
`;
